controllers/bmcjob_controller.go

```diff
@@ -18,45 +18,249 @@ package controllers
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the BMCJob object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+// Reconcile runs a BMCJob.
+// Creates the individual BMCTasks on the cluster.
+// Watches for BMCTask and creates next BMCJob Task based on conditions.
 func (r *BMCJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := r.logger.WithValues("BMCJob", req.NamespacedName)
+	logger.Info("Reconciling BMCJob")
+
+	// Fetch the bmcJob object
+	bmcJob := &bmcv1alpha1.BMCJob{}
+	err := r.client.Get(ctx, req.NamespacedName, bmcJob)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Failed to get BMCJob")
+		return ctrl.Result{}, err
+	}
+
+	// Deletion is a noop.
+	if !bmcJob.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	// Job is Completed or Failed is noop.
+	if bmcJob.HasCondition(bmcv1alpha1.JobCompleted, bmcv1alpha1.ConditionTrue) ||
+		bmcJob.HasCondition(bmcv1alpha1.JobFailed, bmcv1alpha1.ConditionTrue) {
+		return ctrl.Result{}, nil
+	}
+
+	// Create a patch from the initial BMCJob object
+	// Patch is used to update Status after reconciliation
+	bmcJobPatch := client.MergeFrom(bmcJob.DeepCopy())
+
+	// Check if Job is not currently Running
+	if !bmcJob.HasCondition(bmcv1alpha1.JobRunning, bmcv1alpha1.ConditionTrue) {
+		// Initialize the StartTime for the Job
+		now := metav1.Now()
+		bmcJob.Status.StartTime = &now
+		// Set the Job to Running
+		bmcJob.SetCondition(bmcv1alpha1.JobRunning, bmcv1alpha1.ConditionTrue)
+		err = r.patchStatus(ctx, bmcJob, bmcJobPatch)
+		return ctrl.Result{}, err
+	}
+
+	return r.reconcile(ctx, bmcJob, bmcJobPatch, logger)
+}
+
+func (r *BMCJobReconciler) reconcile(ctx context.Context, bmj *bmcv1alpha1.BMCJob, bmjPatch client.Patch, logger logr.Logger) (ctrl.Result, error) {
+	// Get BaseboardManagement object for the Job
+	// Requeue if error
+	baseboardManagement := &bmcv1alpha1.BaseboardManagement{}
+	err := r.getBaseboardManagement(ctx, bmj.Spec.BaseboardManagementRef, baseboardManagement)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, fmt.Errorf("get BMCJob %s/%s BaseboardManagementRef: %v", bmj.Namespace, bmj.Name, err)
+	}
+
+	// List all BMCTask owned by BMCJob
+	bmcTasks := &bmcv1alpha1.BMCTaskList{}
+	err = r.client.List(ctx, bmcTasks, client.MatchingFields{jobOwnerKey: bmj.Name})
+	if err != nil {
+		return ctrl.Result{Requeue: true}, fmt.Errorf("failed to list owned BMCTasks for BMCJob %s/%s", bmj.Namespace, bmj.Name)
```

Comment:
The `Requeue` is unnecessary. Returning an error invokes immediate re-queue behavior.